change the AdvancementProvider to make it possabible for mods to hook into it

Unbekannt1998 opened the pull request:
Since all common Data Providers have a way for mods to hook into, except for the `AdvancementProvider`. So i changed the `AdvancementProvider` so other Mods can use it to make there own Advancement. I also needed to hook in the `Advancement.Builder` and  add a way to give a `ResourceLocation` instead of just a string, since all new Advancements would be under the minecraft namespace. 

I included a Test Mod which create 2 new Advancements and overrides one of Minecrafts default Advancements.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=7488) <br/>All committers have signed the CLA.

ChampionAsh5357 commented:
This should be using `ExistingFileHelper` to determine whether the parents are present before construction. As clarification, `ExistingFileHelper` should be used whenever you are trying to access existing json files that are included with the generator. In this case, since advancements have a parent, they should query the helper to verify that the file exists before continuing to create the advancement json. Since root files also have the ability to not have a parent, you would need to check that the parent tag is not null before verifying the existence using the helper.

Cyborgmas commented on patches/minecraft/net/minecraft/advancements/Advancement.java.patch:
```diff
@@ -1,6 +1,23 @@
+-      public Advancement func_203904_a(Consumer<Advancement> p_203904_1_, String p_203904_2_) {
+-         Advancement advancement = this.func_192056_a(new ResourceLocation(p_203904_2_));
+-         p_203904_1_.accept(advancement);
++      public Advancement func_203904_a(Consumer<Advancement> p_203904_1_, String p_203904_2_){
```
YOu changed a space here after between the ){ creating 2 patch lines. But I would also suggest to not modify the constructor and simply add the other. This would actually be less patch lines overall in this case.

Cyborgmas commented on src/test/java/net/minecraftforge/debug/DataGeneratorTest.java:
```diff
@@ -136,54 +153,55 @@ protected void registerRecipes(Consumer<IFinishedRecipe> consumer)
             ResourceLocation ID = new ResourceLocation("data_gen_test", "conditional");
 
             ConditionalRecipe.builder()
```
Maybe im reading this wrong but it seems to be a lot of unecessary line changes below, changing less lines helps better understand what you did and conserves blame.

Cyborgmas commented on src/test/java/net/minecraftforge/debug/DataGeneratorTest.java:
```diff
@@ -32,28 +32,28 @@
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
-import net.minecraft.block.*;
```
The imports got rearranged which is unecessary diff if you could change that.

Cyborgmas commented on src/test/java/net/minecraftforge/debug/DataGeneratorTest.java:
```diff
@@ -136,54 +153,55 @@ protected void registerRecipes(Consumer<IFinishedRecipe> consumer)
-                                        itemExists("minecraft", "dirt"),
-                                        FALSE()
-                                )
+                                    and(
```
Again here and below you add some tabs, might be better/more correct but its still unecessary.

Cyborgmas commented on src/test/java/net/minecraftforge/debug/DataGeneratorTest.java:
```diff
@@ -435,7 +457,8 @@ protected void registerStatesAndModels()
 
             ModelFile barrel = models().cubeBottomTop("barrel", mcLoc("block/barrel_side"), mcLoc("block/barrel_bottom"), mcLoc("block/barrel_top"));
-            ModelFile barrelOpen = models().cubeBottomTop("barrel_open", mcLoc("block/barrel_side"), mcLoc("block/barrel_bottom"), mcLoc("block/barrel_top_open"));
+            ModelFile barrelOpen = models()
```
No need to change this

Cyborgmas reviewed (commented):
I think it would be better if you revert your style changes to `DataGeneratorTest`, it leads to unecessary diff and makes your changes seem a lot bigger then what they are.

Unbekannt1998 committed:
added a check if a provided RL exists

Unbekannt1998 commented:
i added the check for if the RL is a existing file. 
also changed the Formatting of the test file

TheCurle commented on patches/minecraft/net/minecraft/data/AdvancementProvider.java.patch:
```diff
@@ -0,0 +1,38 @@
+ 
++      registerAdvancement(consumer, fileHelper);
++
++      if(this.getClass() == AdvancementProvider.class)//Forge: Subclasses don't need this.
```
Are you 100% certain that this won't ever become necessary for other mods?

Unbekannt1998 replied:
yes since in the loop only vanilla Advancements will be generated and mods should use `registerAdvancement` which in essence does the same as the loop.

TheCurle replied:
This should be evident from the code, not from looking at this conversation :)
", use registerAdvancement instead" would suffice

TheCurle commented on patches/minecraft/net/minecraft/data/AdvancementProvider.java.patch:
```diff
@@ -0,0 +1,38 @@
+    }
+ 
++   protected void registerAdvancement(Consumer<Advancement> consumer, net.minecraftforge.common.data.ExistingFileHelper fileHelper){
++
```
Is this supposed to be empty?

TheCurle replied:
I understand this is for subclasses to override, but seems like you could at least put a javadoc or something here explaining that, as someone else reading this may not understand it.

Unbekannt1998 replied:
yes it should be empty since all vanilla Advancements are in a list and don't need to be generated again since they would be generated in the loop in the method above. Also I can't make the function abstract. 
I will add a javadoc to it any specifics i should add?

LexManos replied:
Better option is to do what I did with the old version of the recipe provider.
And just move that for loop into this `accept(Consumer<Advancmenet>)` function.
So it's obvious that you should override the function if you want to not have the vanilla stuff.
This also fixes the annoying getClass() test you have to noop the vanilla stuff.

TheCurle commented on src/test/java/net/minecraftforge/debug/DataGeneratorTest.java:
```diff
@@ -32,7 +32,9 @@
 
+import net.minecraft.advancements.criterion.InventoryChangeTrigger;
 import net.minecraft.block.*;
+import net.minecraft.data.*;
```
Did you add this?
Try to avoid wildcard imports.

Unbekannt1998 replied:
Intellij does this automatically for me i will change this for forge

TheCurle reviewed (commented):
Mostly some questions to further understanding of what's going on here.
Some design decisions don't make much sense to me, but you have a chance to explain yourself :)
Overall, though, seems good.

Unbekannt1998 committed:
added a javaDoc for registerAdvancement Method
removed the wildcard import

sciwhiz12 commented on patches/minecraft/net/minecraft/advancements/Advancement.java.patch:
```diff
@@ -1,6 +1,43 @@
+          }
+       }
+ 
++      @Deprecated //Forge: use the other Method with the fileHelper
```
Maybe `//Forge: Use Advancement#register(Consumer<Advancement>, ResourceLocation, ExistingFileHelper)`?

sciwhiz12 commented on patches/minecraft/net/minecraft/data/AdvancementProvider.java.patch:
```diff
@@ -0,0 +1,44 @@
+ 
+    }
+ 
++   /**
++    * Subclasses should override this function if they want to generate Advancements.
++    * Just use {@see Advancement.Builder} to build your Advancements, you don't need an extra consumer like the vanilla classes.
++    * @param consumer used for the build function from {@see Advancement.Builder}
++    * @param fileHelper used for the build function from {@see Advancement.Builder}
++    */
++   protected void registerAdvancement(Consumer<Advancement> consumer, net.minecraftforge.common.data.ExistingFileHelper fileHelper){
```
I would suggest maybe a rename to `registerAdvancements`, to be consistent with the other similarly-functioning methods in datagen? (`registerTags` for `TagsProvider`, `registerRecipes` in `RecipeProvider`, `registerStatesAndModels` in `BlockStateProvider`, etc.)

Also, a suggestion for the javadoc:
```java
/**
 * Override this method for registering and generating custom {@link Advancement}s.
 * @param consumer Consumer for registering advancements to this provider
 * @param fileHelper Existing file helper
 * @see Advancement.Builder#build(Consumer<Advancement>, ResourceLocation, net.minecraftforge.common.data.ExistingFileHelper)
 */
```

Unbekannt1998 replied:
i would like to keep the second part of my javadoc, since it is more clear what you should use, because if someone looks at the vanilla Classes they could be confused with the additional  consumer they use. Also Intellij does not show the last line when you look at the javadoc on the overridden Method.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/DataGeneratorTest.java:
```diff
@@ -645,6 +649,77 @@ public String getName() {
+                    .withCriterion("crafting_table", InventoryChangeTrigger.Instance.forItems(Blocks.CRAFTING_TABLE))
+                    .register(consumer, new ResourceLocation("story/root"), fileHelper);
+
+/*            Advancement.Builder.builder().withDisplay(Blocks.COBBLESTONE,
+                    new TranslationTextComponent(Items.COBBLESTONE.getTranslationKey()),
+                    new StringTextComponent("You got cobblestone"),
+                    new ResourceLocation("textures/gui/advancements/backgrounds/stone.png"),
+                    FrameType.TASK,
+                    false,
+                    false,
+                    false)
+                    .withCriterion("get_cobbleStone", InventoryChangeTrigger.Instance.forItems(Items.COBBLESTONE))
+                    .withParentId(new ResourceLocation("not_there/not_here"))
+                    .register(consumer, "illegalParent");*/
```
From the advancement's name, I assume this is for testing if a exception will happen when the parent is not present/does not exist? 
Please add a comment to clarify it (`// This should cause an error because of the parent not existing`), and also pass in the `ExistingFileHelper` to the `register(consumer, "illegalParent")` call to actually test if the `ExistingFileHelper` does complain about the missing parent.

sciwhiz12 reviewed (changes requested):
A few things I'd like to see changed; the big one being the method rename from `registerAdvancement` to `registerAdvancements`, to be consistent with other data providers.

Unbekannt1998 committed:
rename registerAdvancement Method to registerAdvancements
change the java doc of registerAdvancements slightly
make the commented Test use the new Method and add an explanation why it is commented and what it does
make the comment on the deprecation more clear on what Method to use

Unbekannt1998 commented:
can i get a triage member to look at this and maybe get some feedback on what i still need to change, if needed.

Unbekannt1998 committed:
remove inconsistent spacing in patch also update Forge version

LexManos commented on patches/minecraft/net/minecraft/advancements/Advancement.java.patch:
```diff
@@ -1,6 +1,42 @@
++
++      public Advancement build(ResourceLocation id, net.minecraftforge.common.data.ExistingFileHelper fileHelper) {
++         if (!this.func_192058_a((p_199750_0_) -> {
++            if(fileHelper.exists(p_199750_0_, net.minecraft.resources.ResourcePackType.SERVER_DATA, ".json", "advancements")) {
```
This is too large for a patch. And it makes the code really hard to understand as its in SRG names.
You should move this to a ExistingAdvancementBuilder or something in the forge util package.

Unbekannt1998 replied:
i hope my latest commit is what you meant by this

LexManos commented on patches/minecraft/net/minecraft/data/AdvancementProvider.java.patch:
```diff
@@ -0,0 +1,44 @@
+       this.field_204025_c = p_i48869_1_;
+    }
+ 
++   public AdvancementProvider(DataGenerator generatorIn, net.minecraftforge.common.data.ExistingFileHelper fileHelperIn){ this.field_204025_c = generatorIn; this.fileHelper = fileHelperIn;}
```
Don't do crap like this.
It makes the line really long and hard to read.

LexManos commented on patches/minecraft/net/minecraft/data/AdvancementProvider.java.patch:
```diff
@@ -0,0 +1,44 @@
+    private static final Gson field_204024_b = (new GsonBuilder()).setPrettyPrinting().create();
+    private final DataGenerator field_204025_c;
+    private final List<Consumer<Consumer<Advancement>>> field_204283_d = ImmutableList.of(new EndAdvancements(), new HusbandryAdvancements(), new AdventureAdvancements(), new NetherAdvancements(), new StoryAdvancements());
++   private net.minecraftforge.common.data.ExistingFileHelper fileHelper;
```
Should be protected to allow the child class to access it.

Unbekannt1998 committed:
Address Comments from Lex

covers1624 commented on patches/minecraft/net/minecraft/advancements/Advancement.java.patch:
```diff
@@ -1,6 +1,37 @@
++      }
++
++      public Advancement build(ResourceLocation id, net.minecraftforge.common.data.ExistingFileHelper fileHelper) {
++         if (!net.minecraftforge.common.data.AdvancementBuilderHelper.build(fileHelper, this)){
```
Space before ending brace. `)){` -> `)) {`

covers1624 commented on patches/minecraft/net/minecraft/data/AdvancementProvider.java.patch:
```diff
@@ -0,0 +1,38 @@
+       this.field_204025_c = p_i48869_1_;
+    }
+ 
++   public AdvancementProvider(DataGenerator generatorIn, net.minecraftforge.common.data.ExistingFileHelper fileHelperIn){
```
Again space before brace.

covers1624 commented on patches/minecraft/net/minecraft/data/AdvancementProvider.java.patch:
```diff
@@ -0,0 +1,38 @@
++    * @param consumer used for the register function from {@link Advancement.Builder}
++    * @param fileHelper used for the register function from {@link Advancement.Builder}
++    */
++   protected void registerAdvancements(Consumer<Advancement> consumer, net.minecraftforge.common.data.ExistingFileHelper fileHelper){
```
Again space before brace.

covers1624 commented on src/main/java/net/minecraftforge/common/data/AdvancementBuilderHelper.java:
```diff
@@ -0,0 +1,18 @@
+package net.minecraftforge.common.data;
```
License header. Run the `updateLicense` Gradle task.

covers1624 commented on src/main/java/net/minecraftforge/common/data/AdvancementBuilderHelper.java:
```diff
@@ -0,0 +1,18 @@
+
+public class AdvancementBuilderHelper {
+
+    public static boolean build(ExistingFileHelper fileHelper, Advancement.Builder builder){
```
Space before brace.

covers1624 commented on src/main/java/net/minecraftforge/common/data/AdvancementBuilderHelper.java:
```diff
@@ -0,0 +1,18 @@
+
+    public static boolean build(ExistingFileHelper fileHelper, Advancement.Builder builder){
+        return builder.resolveParent((advancementId) -> {
+            if(fileHelper.exists(advancementId, ResourcePackType.SERVER_DATA, ".json", "advancements")) {
```
space after `if` before the first parenthesis.

Unbekannt1998 committed:
Address covers comments
added License Header and some spaces

Unbekannt1998 committed:
update to current forge

Unbekannt1998 commented:
updated to the last Forge version and changed the names to the Mojang Mapping names

Unbekannt1998 committed:
add missing patch for AdvancementProvider i forgot for some reason

Unbekannt1998 committed:
update License Header

Unbekannt1998 committed:
added spaces before the braces

ChampionAsh5357 commented on patches/minecraft/net/minecraft/advancements/Advancement.java.patch:
```diff
@@ -1,5 +1,36 @@
+          return advancement;
+       }
+ 
++      public Advancement save(Consumer<Advancement> consumer, ResourceLocation id, net.minecraftforge.common.data.ExistingFileHelper fileHelper) {
++         Advancement advancement = this.build(id, fileHelper);
++         consumer.accept(advancement);
++         return advancement;
++      }
++
++      public Advancement build(ResourceLocation id, net.minecraftforge.common.data.ExistingFileHelper fileHelper) {
++         if(!net.minecraftforge.common.data.AdvancementBuilderHelper.build(fileHelper, this)) {
++            throw new IllegalStateException("Tried to build incomplete advancement!");
++         } else {
++            if (this.field_192066_f == null) {
++               this.field_192066_f = this.field_199751_g.createRequirements(this.field_192065_e.keySet());
++            }
++            return new Advancement(id, this.field_192062_b, this.field_192063_c, this.field_192064_d, this.field_192065_e, this.field_192066_f);
++         }
++      }
```
So I'm going to reiterate Lex's point here. Most of this could be moved into a different class. Technically, all you should need to do here is have a method that takes in the advancement builder, data provider consumer, id, and the existing file helper. You could then call `#canBuild` within that method with the function to create a dummy advancement (which is what you are doing in `AdvancementBuilderHelper#build` and then `Builder#build` with no issues. Most you should need is three lines here, one for the method definition, one for the return statement to the outside method call, and one for the end bracket.

ChampionAsh5357 commented on patches/minecraft/net/minecraft/data/AdvancementProvider.java.patch:
```diff
@@ -0,0 +1,38 @@
++   }
++
++   /**
++    * Override this method for registering and generating custom {@link Advancement}s. <br>
```
I believe javadoc convention is to use an opening `<p>` on line breaks.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/data/AdvancementBuilderHelper.java:
```diff
@@ -0,0 +1,37 @@
+public class AdvancementBuilderHelper {
+
+    public static boolean build (ExistingFileHelper fileHelper, Advancement.Builder builder) {
+        return builder.canBuild((advancementId) -> {
+            if (fileHelper.exists(advancementId, ResourcePackType.SERVER_DATA, ".json", "advancements")) {
+                return new Advancement(advancementId, null, null, AdvancementRewards.EMPTY, Maps.newHashMap(), null);
+            }
+            return null;
+        });
```
When talking about the other comment. This is pretty much what I mean since you can do this and build the advancement in one method. You can ever return a more detailed error message specific to data providers.

ChampionAsh5357 reviewed (changes requested):
Mainly just need to lower the patches to `Advancement`, rest seems pretty good.

Unbekannt1998 committed:
reduce Patchsize
update testFile to Mojang Mappings

ChampionAsh5357 commented on patches/minecraft/net/minecraft/data/AdvancementProvider.java.patch:
```diff
@@ -0,0 +1,39 @@
++
++   /**
++    * Override this method for registering and generating custom {@link Advancement}s.
++    * <p>
++    * Just use {@link Advancement.Builder} to build your Advancements, you don't need an extra consumer like the vanilla classes.
```
You can move them to the same line.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/data/AdvancementBuilderHelper.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.Map;
+import java.util.function.Consumer;
+
+public class AdvancementBuilderHelper {
```
Braces on new line for everything in the class.

ChampionAsh5357 reviewed (commented):
I'll test this when I get home in a few hours to verify everything works. For now, just minor formatting fixes.

Unbekannt1998 committed:
style changes

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/data/AdvancementBuilderHelper.java:
```diff
@@ -0,0 +1,51 @@
+    {
+        boolean canBuild = builder.canBuild((advancementId) ->
+        {
+            if (fileHelper.exists(advancementId, ResourcePackType.SERVER_DATA, ".json", "advancements")) {
```
Rest of this looks good to me. Just this one brace on new line.

ChampionAsh5357 commented:
Probably can remove `.cache` and `good_parent` since they are both changes during datagen.

Unbekannt1998 committed:
move brace to next line
remove good_parent and remove cache info

Unbekannt1998 committed:
remove space between funktion name and parameters

Unbekannt1998 committed:
back Port changes from #8120
